compiler/src/dotty/tools/dotc/transform/IsInstanceOfEvaluator.scala

```diff
@@ -147,7 +185,7 @@ class IsInstanceOfEvaluator extends MiniPhaseTransform { thisTransformer =>
 
           val inMatch = s.qualifier.symbol is Case
```

Comment:
The `caseIsInstanceOf` idea looks right (exclusive or not: no idea, propose something :-) It could be a separate PR.